src/librustdoc/lib.rs

```diff
@@ -109,7 +142,25 @@ fn get_args() -> Option<Vec<String>> {
                 })
                 .ok()
         })
-        .collect()
+        .collect::<Option<_>>()
+        .map(|mut args: Vec<String>| {
+            // Make sure we use the right default sysroot. The default sysroot is wrong,
+            // because `get_or_default_sysroot` in `librustc_session` bases that on `current_exe`.
+            //
+            // Make sure we always call `compile_time_sysroot` as that also does some sanity-checks
+            // of the environment we were built in.
+            // FIXME: Ideally we'd turn a bad build env into a compile-time error via CTFE or so.
+            if let Some(sysroot) = compile_time_sysroot() {
+                let sysroot_flag = "--sysroot";
+                if !args.iter().any(|e| e == sysroot_flag) {
+                    // We need to overwrite the default that librustc_session would compute.
+                    args.push(sysroot_flag.to_owned());
+                    args.push(sysroot);
+                }
+            }
+
+            args
```

Comment:
I'm not 100% sure that it's the most readable way to update the `Vec<String>` in this `Option`.